### PR TITLE
Attempt at avoiding falsely flagged buffer overruns

### PR DESCRIPTION
**Pull request by erikbern**
Just a quick shot at addressing https://github.com/jlmelville/uwot/issues/50

### CODE REVIEW

File: src/annoylib.h [hunk 1]
```diff
@@ -1142,19 +1146,25 @@ template<typename S, typename T, typename Distance, typename Random>
     if (n > _nodes_size) {
       const double reallocation_factor = 1.3;
       S new_nodes_size = std::max(n, (S) ((_nodes_size + 1) * reallocation_factor));
-      void *old = _nodes;
+      void *new_nodes = NULL;
```
Comment (erikbern) [inline src/annoylib.h]: oops this was something that was irrelevant to this PR, let me revert

**Comment by eddelbuettel** [comment]
Could we ever run over MAX_ARRAY_SIZE ?

**Comment by eddelbuettel** [comment]
Oh, gosh, I had though `MAX_ARRAY_SIZE` had been defined. Woulnd't 

```c
#define MAX_ARRAY_SIZE (1<<30)
```

together with 

```
T v[MAX_ARRAY_SIZE];
```

put massive arrays on the stack?  

Confused...

**Comment by erikbern** [comment]
> put massive arrays on the stack?

No because the Node object is actually never allocated if you read the code. So it should be fine!

**Comment by eddelbuettel** [comment]
Testing now at rhub.  Old code gave 

```
   ../inst/include/annoylib.h:364:24: runtime error: index 1 out of bounds for type 'float [1]'         
   ../inst/include/annoylib.h:364:46: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:364:13: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:359:24: runtime error: index 1 out of bounds for type 'float [1]'          
   ../inst/include/annoylib.h:359:46: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:359:13: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:736:21: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:736:31: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:736:11: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:394:9: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                     
   ../inst/include/annoylib.h:740:20: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:740:31: runtime error: index 1 out of bounds for type 'float [1]'                                                                                                                    
   ../inst/include/annoylib.h:740:41: runtime error: index 1 out of bounds for type 'float [1]'   
```

Will know in a moment if this helps.  Not a perfect test (as we don't have the exact setup CRAN uses).

**Comment by eddelbuettel** [comment]
Part of me has the vague feeling that we once had a constant there, also got yelled at, and later went to `v[1]`.  But that was a while back and it is all a little fuzzy...

**Comment by erikbern** [comment]
> Part of me has the vague feeling that we once had a constant there, also got yelled at, and later went to `v[1]`. But that was a while back and it is all a little fuzzy...

I think it was zero for a long time. It's like the old saying that in computer science there's only three numbers, 0, 1, and n. And we went from 0 to 1 to now n :)

**Comment by eddelbuettel** [comment]
Sadly it didn't yet work.  
```
   ../inst/include/annoylib.h:368:24: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:368:46: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:368:13: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:363:24: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:363:46: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:363:13: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:740:21: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:740:31: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:740:11: runtime error: index 1 out of bounds for type 'float [1]'          
   ../inst/include/annoylib.h:398:9: runtime error: index 1 out of bounds for type 'float [1]'
   ../inst/include/annoylib.h:744:20: runtime error: index 1 out of bounds for type 'float [1]'          
   ../inst/include/annoylib.h:744:31: runtime error: index 1 out of bounds for type 'float [1]'          
   ../inst/include/annoylib.h:744:41: runtime error: index 1 out of bounds for type 'float [1]' 
```

**Comment by jlmelville** [comment]
Is it possible that the declaration needs to be changed in more places, e.g.: 

https://github.com/spotify/annoy/blob/d6dc8050febb6c6e82353a85770bf5e43cb9778f/src/annoylib.h#L517

**Comment by jlmelville** [comment]
If I replace all declarations of `T v[1]` with `T [V_ARRAY_SIZE]` in `annoylib.h`, then when running the GCC UBSAN checker via `rhub::check_with_sanitizers()` with the RcppAnnoy project, I no longer see the `../inst/include/annoylib.h:872:11: runtime error: index 1 out of bounds for type 'float [1]'` errors. For reference in the new version of `annoylib.h`, that line is:

https://github.com/spotify/annoy/blob/d6dc8050febb6c6e82353a85770bf5e43cb9778f/src/annoylib.h#L897

However, I now see a new runtime error:

```
../inst/include/annoylib.h:662:18: runtime error: load of misaligned address 0x55d2429c5424 for type 'const long unsigned int', which requires 8 byte alignment
0x55d2429c5424: note: pointer points here
  00 00 00 00 00 00 00 00  00 00 00 00 01 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
```

the line referenced is:

https://github.com/spotify/annoy/blob/d6dc8050febb6c6e82353a85770bf5e43cb9778f/src/annoylib.h#L662

**Comment by eddelbuettel** [comment]
Good thinking about replacing the other assignments!  I was rushing things before making dinner.

@erikbern Any magic for that line 662?

**Comment by erikbern** [comment]
Turns out there's actually 4 places this had to be modified, and I had just fixed it in one of them.

Can you check if it works now? Thanks!

**Comment by jlmelville** [comment]
Sorry for any confusion @erikbern, my previous comment was for what is now change 81648fb, i.e. I had changed those four locations already.

As of <https://github.com/spotify/annoy/pull/455/commits/81648fbc26f7c1b924affc9960acf6191a637115>, the old error has disappeared, but a new one has appeared, related to line 662.

**Comment by erikbern** [comment]
> ../inst/include/annoylib.h:662:18: runtime error: load of misaligned address 0x55d2429c5424 for type 'const long unsigned int', which requires 8 byte alignment

Actually this PR won't fix that issue, and I'm not sure how to even fix it. Some thoughts:
* Make sure that the memory is always aligned to 8 byte offsets, possibly using something like http://man7.org/linux/man-pages/man3/posix_memalign.3.html
* Rewrite the hamming code to use 4 byte ints instead of 8 byte ints
* Change the few places where we access `v[i]` to not use an array dereference

None of these are entirely trivial

**Comment by jlmelville** [comment]
Fair enough. I would be happy to see the current PR merged, as it does localize the UBSAN complaints to just the Hamming calculation, which is an improvement.

I'd like to open a separate issue to cover the new misalignment problem, even if there isn't an obvious fix.

**Comment by erikbern** [comment]
Cool will handle the Hamming issue in a separate PR, thanks

**Comment by jlmelville** [comment]
Thanks for your work on this @erikbern, it is much appreciated.

**Comment by eddelbuettel** [comment]
Very much seconded.  That was excellent help, and really timely :)

**Pull request merged**